src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -307,17 +319,43 @@
      *
      * @param repoOwner the repository owner.
      * @param repository the repository name.
+     * @param repositoryUrl HTML URL for the repository. If specified, takes precedence over repoOwner and repository.
+     * @param configuredByUrl Whether to use repositoryUrl or repoOwner/repository for configuration.
+     * @throws IllegalArgumentException if repositoryUrl is specified but invalid.
      * @since 2.2.0
      */
+    // configuredByUrl is used to decide which radioBlock in the UI the user had selected when they submitted the form.
     @DataBoundConstructor
-    public GitHubSCMSource(@NonNull String repoOwner, @NonNull String repository) {
-        this.repoOwner = repoOwner;
-        this.repository = repository;
+    public GitHubSCMSource(String repoOwner, String repository, String repositoryUrl, boolean configuredByUrl) {
```

Comment:
I'm struggling a little with this PR. My understanding is that all of this `deprecated` stuff is "short-term", but I can't help but feel that "short-term" is actually "long-term" unless there's a new major version coming soon...

The problem arose because this change caused an error in our JobDSL execution _(I wrote about that here, but I think that might a different problem: https://groups.google.com/forum/?pli=1#!topic/job-dsl-plugin/UnnGPBTlgA8)_

What bothers me is that the previous constructor has been deprecated, however it seems to me like `configuredByUrl` is optional. And if `configuredByUrl` is false, then `repositoryUrl` isn't used at all (meaning it is also optional). So why would the previous constructor be deprecated? Or at least, why is there a new constructor that mixes deprecated functionality with the new functionality? 

----
The other thing I don't follow, is that in the new constructor, we check if `configuredByUrl` is set:
* If it is not set, then `repoOwner` & `repository` must not be null, however there's no check to make sure they're not (the old constructor used the `@NonNull` annotation, and the deprecated one took that away)
* If it is set, than `repositoryUrl` must not be null  _(well `forRepositoryUrl() has a check for @NonNull so maybe that's actually enough, I'm not sure.)_